src/linear_algebra/annihilating_polynomial.lean

```diff
@@ -0,0 +1,139 @@
+as defined in `field_theory.minpoly`.
+
+## Special case: endomorphism algebra
+
+Given an `R`-module `M` (`[add_comm_group M] [module R M]`)
+there are some common specialiazations which may be more familiar.
+* Example 1: `A = M →ₗ[R] M`, the endomorphism algebra of an `R`-module M.
+* Example 2: `A = n × n` matrices with entries in `R`.
+-/
+
+open_locale polynomial
+
+namespace polynomial
+
+section general_case
+
+variables {R A : Type*} [comm_semiring R] [semiring A] [algebra R A]
+
+variables (R)
+
+/--
+The informal notation `p(a)` stand for `polynomial.aeval a p`.
+Again informally, the annihilating ideal of `a` is
+`{ p ∈ R[x] | p(a) = 0 }`. This is an ideal in `R[X]`.
+The formal definition uses the kernel of the aeval map. -/
+noncomputable def annihilating_ideal (a : A) : ideal R[X] :=
+ring_hom.ker (aeval a).to_ring_hom
+
+variables {R}
+
+/-- It is useful to refer to ideal membership sometimes
+ and the annihilation condition other times -/
+lemma mem_annihilating_ideal_iff_aeval_0 (a : A) (p : R[X]) :
+  p ∈ annihilating_ideal R a ↔ aeval a p = 0 :=
+iff.rfl
+
+end general_case
+
+section field_case
+
+variables {𝕜 A : Type*} [field 𝕜] [ring A] [algebra 𝕜 A]
+variables (𝕜)
+
+/-- Since `𝕜[x]` is a principal ideal domain there is a polynomial `g` such that
+ `span 𝕜 {g} = annihilating_ideal a` -/
+noncomputable def annihilating_ideal_generator (a : A) : 𝕜[X] :=
+submodule.is_principal.generator (annihilating_ideal 𝕜 a)
+
+variables {𝕜}
+
+section minpoly_generates
+
+/-- We are working toward showing the generator of the annihilating ideal
+in the field case is the minimal polynomial. We are going to use a uniqueness
+theorem of the minimal polynomial. This is the first condition: it must annihilate
+the original element `a : A`. -/
+lemma annihilating_ideal_generator_aeval_0 (a : A) :
+  aeval a (annihilating_ideal_generator 𝕜 a) = 0 :=
+begin
+  rw annihilating_ideal_generator,
+  have gen_member := submodule.is_principal.generator_mem (annihilating_ideal 𝕜 a),
+  exact (ring_hom.mem_ker (polynomial.aeval a).to_ring_hom).1 gen_member,
+end
+
+/-- This is a stepping stone to show the generator has minimal degree -/
+lemma mem_iff_generator_dvd (a : A) (p : 𝕜[X]) :
+  p ∈ annihilating_ideal 𝕜 a ↔ annihilating_ideal_generator 𝕜 a ∣ p :=
+submodule.is_principal.mem_iff_generator_dvd (annihilating_ideal 𝕜 a)
+
+/-- The generator of the annihilating ideal has minimal degree among
+ the non-zero members of the annihilating ideal -/
+lemma degree_annihilating_ideal_generator_le_of_mem (a : A) (p : 𝕜[X])
+  (hp : p ∈ annihilating_ideal 𝕜 a) (hpn0 : p ≠ 0) :
+  degree (annihilating_ideal_generator 𝕜 a) ≤ degree p :=
+degree_le_of_dvd ((mem_iff_generator_dvd a p).1 hp) hpn0
+
+/-- This is what we have been building to:
+The monic generator of the annihilating ideal is the minimal polynomial. -/
+lemma minpoly_eq_monic_annihilating_ideal_generator (a : A)
+  (h : (annihilating_ideal_generator 𝕜 a).monic) :
```

Comment:
as you're in a field, can't you try normalize the generator in the definition so that it's always monic? I feel like this is the case that people will want the most.

Reply:
`polynomial.leading_coeff` is probably the declaration you'd be looking for to do that

Reply:
I am struggling with the syntax of the if/else. Taking 0 when the annihilating ideal is 0/taking g/g.leading_coeff otherwise. I am looking at similar things in the minpoly.lean hoping it clicks ...

Reply:
I don't think you need an if-else - if you want inspiration and no spoilers, you can look at `monic.lean`, but if you want a spoiler:

<details>
  <summary>soln</summary>
you can just multiply by `(C (leading_coeff)) ^ -1`, as we're in a field
</details>

Reply:
getting into lean details below. Should I move this question to Zulip chat?

1. I am surprised g.leading_coeff works. I don't understand why lean does not complain about the fact that g.leading_coeff might be 0

2. I went about re-proving everything for the monic generator case. But I can't help but feel this is not the right way. For example, proofs that use facts about submodule.is_principal.generator don't work for me because I took that definition and scaled the generator it gave me to make it monic. There must be some way to convince lean that all those proofs work for any choice of generator and that my definition is one such choice.

Reply:
For 1, you'll want to see this: https://xenaproject.wordpress.com/2020/07/05/division-by-zero-in-type-theory-a-faq/ and note that this coincidentally works out when the leading coeff is zero!

For 2, this is sadly the cost of making a new defn - and because submodule.is_principal is defined using an existential, there's no real way to reuse the API, and you can only really use the lemmata about span singleton. If someone made API for generating ideals, you could prove that this is an ideal that generates and reuse that.

Reply:
> If someone made API for generating ideals
Does `API` mean `typeclass plus some theorems about it` ?

Reply:
apologies! I didn't see this message till now, it got lost in my emails. It just means essentially lemmas such that working with the definition doesn't require unfolding it every time, or going in the nitty gritty; I wouldn't want to have to refer to how `int` is defined in order to show -32 < 234, for example.

Reply:
@ericrbg I finally got around to rewriting this with monic in the definition. This brought way more difficulties than I thought it would. What I have now works, but is not very nice. I copied over lemmas from is_principal.generator in ring_theory/principal_ideal_domain.lean.  It would be better to be able to use lemmas that work for any choice of generator. I am using the current proofs to guide me on what sorts of lemmas would be helpful. I am open to suggestions.

Reply:
I'm afraid you're really overcomplicating the matters - I'll try review more carefully when I get some time, but for example:

```lean
lemma span_singleton_ann_ideal_generator (a : A) :
  span 𝕜[X] {ann_ideal_generator 𝕜 a} = ann_ideal 𝕜 a :=
begin
  simp only [ann_ideal_generator, ann_ideal, alg_hom.to_ring_hom_eq_coe, ideal.submodule_span_eq],
  rw ideal.span_singleton_mul_left_unit,
  { exact is_principal.span_singleton_generator _ },
  { rw [is_unit_C, is_unit_iff_ne_zero],
    apply inv_ne_zero,
    rw [ne.def, leading_coeff_eq_zero_iff_deg_eq_bot] },
end
```

and all you have to deal with is the case that the ideal is zero, which doesn't seem to be too hard.